classes/Helper.php

```diff
@@ -0,0 +1,122 @@
+ */
+
+namespace report_rolessitemap;
+use moodle_url;
+use report_rolessitemap\Misc;
+
+defined('MOODLE_INTERNAL') || die();
+
+class Helper {
+    /**
+     * Shows all categories and the assigned users with their roles
+     *
+     * @return void
+     * @throws \dml_exception
+     */
+    public function rendercategoricesandroles(): void {
+        GLOBAL $DB, $OUTPUT;
+        $systemcontext = \context_system::instance();
+        $roles = role_fix_names(get_all_roles(), $systemcontext, ROLENAME_ORIGINAL);
+        //$roles = $DB->get_records('role', null, 'id', 'id, shortname');
+        $categorieslist = \core_course_category::make_categories_list();
+        // Now populate $categorieslistandroles with the information to be rendered using mustach-template.
+        $categorieslistandroles = [];
+        foreach ($categorieslist as $categoryid => $categoryname) {
+            $context = \context_coursecat::instance($categoryid);
+            $params = array('contextid' => $context->id);
+            $roleassignments = $DB->get_records('role_assignments', $params, 'contextid, roleid, userid', 'id, contextid, roleid, userid');
+            $roleassignmentsasarray = [];
+            $oldshortname = "";
+            foreach ($roleassignments as $roleassignment) {
+                $counter = false;
+                if (!in_array($roleassignment->roleid, $this->get_supported_roles())) {
+                    continue;
+                }
+                if ($roles[$roleassignment->roleid]->shortname != $oldshortname) {
+                    $oldshortname = $roles[$roleassignment->roleid]->shortname;
+                    $counter = true;
+                }
+                $roleassignmenteditingurl = new moodle_url('/admin/roles/assign.php',
+                    array('contextid' => $context->id , 'roleid' => $roleassignment->roleid));
+                $userprofilgurl = new moodle_url('/user/profile.php',
+                    array('userid' => $roleassignment->userid));
+                $user = \core_user::get_user($roleassignment->userid,  'username, firstname, lastname');
+                $roleassignmentsasarray[] = [
+                    "counter" => $counter,
+                    "roleid" => $roleassignment->roleid,
+                    'roleassignmenteditingurl' => $roleassignmenteditingurl->out(false),
+                    "shortname" => $roles[$roleassignment->roleid]->shortname,
+                    "localname" => $roles[$roleassignment->roleid]->localname,
+                    "userid" => $roleassignment->userid,
+                    "userprofilgurl" => $userprofilgurl->out(false),
+                    "username" => fullname($user)
+                ];
+            }
+            $url = new moodle_url('/course/index.php', array('categoryid' => $categoryid));
+            $categorieslistandroles[] = [
+                'categoryid' => "$categoryid",
+                'categoryname' => "$categoryname",
```

Comment:
why the quotes?

Reply:
fixed; mayby earlier state of code to be sure o have strings ?